### DISCUSSION TITLE
Error with JSON decoding from Go API

### DISCUSSION THREAD

**Discussion by jgillich** [cue-lang/cue#1477] [Q&A] [answered]
Trying to load JSON from a tag and use it in another package. Here is a small demo: https://github.com/jgillich/cue-sample

`foo` defines a tag and decodes it, `bar` imports `foo`. This works fine when called from the cli:

```
$ cue export ./bar ./foo --inject 'foo={"foo":1}'
{
    "f": {
        "foo": 1
    }
}
{
    "foo": 1
}
```

But the Go code doesn't work:

```
go run foo.go 
foo.go:35: cue: marshal error: f: error in call to encoding/json.Unmarshal: non-concrete value string
exit status 1
```

Any ideas what I'm doing wrong here? Am I missing some configuration?

**Comment by xinau** [comment, marked as answer]
@jgillich `cue export` unifies all values before exporting them. I think the cue equivalent of your current code ([6eebffc](https://github.com/jgillich/cue-sample/blob/6eebffc/foo.go)) would be the same as running `cue export ./bar` which yields a similar error.

```
error in call to encoding/json.Unmarshal: non-concrete value string:
    ./foo/foo.cue:6:1
```

Your code works when unifying all values before marshaling them. I.e.

```diff
diff --git a/foo.go b/foo.go
index 9c332b2..f586743 100644
--- a/foo.go
+++ b/foo.go
@@ -30,7 +30,12 @@ func main() {
                log.Fatal(err)
        }

-       json, err := defs[0].MarshalJSON()
+       def := defs[0]
+       for _, d := range defs[1:] {
+               def = def.Unify(d)
+       }
+
+       json, err := def.MarshalJSON()
        if err != nil {
                log.Fatal(err)
        }
```

I hope this helps.

**Reply by jgillich** [reply]
Thank you so much! Had no idea about unify

**Reply by jgillich** [reply]
Ok actually this isn't quite what I want, the problem is that I now end up with a merged variant of both: `{"f":{"foo":1},"foo":1}`

Whereas the cli returns two different objects. In my real scenario they have conflicting types to I cannot unify them

I assume there is a another operation similar to unify, do you know what it could be @xinau?

**Reply by xinau** [reply]
Sorry. You are totally correct. I misinterpreted it somehow. I'm not aware of a similar operation and will have to lookup the implementation to provide you an answer.

**Reply by jgillich** [reply]
I've read through a good amount of the cli code and as far as I can tell it is essentially identical to what I'm doing, only the cli isn't using context, but I haven't found any places where that would've made a difference.

@mpvl Hope I'm not bothering you but his is probably easy to answer for someone familiar with the code base. Do you know what the issue is here?

**Reply by xinau** [reply]
@jgillich I was able to figure out that the difference is somehow between the deprecated [`cue.Build`](https://pkg.go.dev/cuelang.org/go@v0.4.0/cue#Build) and [`cuecontext.BuildInstances`](https://pkg.go.dev/cuelang.org/go@v0.4.0/cue#Context.BuildInstances). When using `cue.Build` with this code I get the following results 

```diff
diff --git a/foo.go b/foo.go
index 9c332b2..264cc0f 100644
--- a/foo.go
+++ b/foo.go
@@ -3,7 +3,7 @@ package main
 import (
 	"log"
 
-	"cuelang.org/go/cue/cuecontext"
+	"cuelang.org/go/cue"
 	"cuelang.org/go/cue/load"
 )
 
@@ -23,18 +23,14 @@ func main() {
 		}
 	}
 
-	cueContext := cuecontext.New()
-
-	defs, err := cueContext.BuildInstances(instances)
-	if err != nil {
-		log.Fatal(err)
-	}
+	defs := cue.Build(instances)
+	for _, def := range defs {
+		val := def.Value()
+		json, err := val.MarshalJSON()
+		if err != nil {
+			log.Fatal(err)
+		}
 
-	json, err := defs[0].MarshalJSON()
-	if err != nil {
-		log.Fatal(err)
+		log.Printf("json: %v", string(json))
 	}
-
-	log.Printf("json: %v", string(json))
-
 }
```

```console
foo.go:34: json: {"f":{"foo":1}}
foo.go:34: json: {"foo":1}
```

But someone with a bit more insight, will probably be able to provide a "working solution" using `BuildInstances`.

**Reply by jgillich** [reply]
Wow, thanks again! The code paths look very similar, both call `runtime.Build`, but the non-context version also calls [`getImportFromBuild`](https://github.com/cue-lang/cue/blob/34c4f9cc12629593419409c2d24afe0db9d11933/cue/build.go#L140) which might explain this issue? I don't fully understand the code but I believe this is a bug so I'll open an issue.
